File: 2021/cpp/day_4.cpp
//
// Created by trox667 on 04.12.21.
//

#include <cassert>
#include <algorithm>
#include <unordered_set>
#include "day_4.hpp"
#include "utils.hpp"

namespace day4 {

    struct Board {
        Board() : last_marked{0}, completed{false} {};

        void add_row(const std::vector<int> &row) {
            grid.emplace_back(row);
        }

        void mark(const int value) {
            for (auto y = 0; y < 5; ++y) {
                for (auto x = 0; x < 5; ++x) {
                    if (grid[y][x] == value) {
                        int idx = y * 5 + x;
                        marker.emplace(idx);
                        last_marked = value;
                    }
                }
            }
        }

        bool is_completed() {
            for (auto i = 0; i < 5; ++i) {
                // TODO: different way to create the indices
                int idxy0 = i * 5 + 0;
                int idxy1 = i * 5 + 1;
                int idxy2 = i * 5 + 2;
                int idxy3 = i * 5 + 3;
                int idxy4 = i * 5 + 4;

                int idxx0 = 0 * 5 + i;
                int idxx1 = 1 * 5 + i;
                int idxx2 = 2 * 5 + i;
                int idxx3 = 3 * 5 + i;
                int idxx4 = 4 * 5 + i;

                if ((marker.contains(idxy0) && marker.contains(idxy1) &&
                     marker.contains(idxy2) && marker.contains(idxy3) &&
                     marker.contains(idxy4)) ||
                    (marker.contains(idxx0) && marker.contains(idxx1) &&
                     marker.contains(idxx2) && marker.contains(idxx3) &&
                     marker.contains(idxx4))) {
                    completed = true;
                    return true;
                }
            }
            return false;
        }

        int sum_unmarked() {
            int sum = 0;
            for (auto y = 0; y < 5; ++y) {
                for (auto x = 0; x < 5; ++x) {
                    int idx = y * 5 + x;
                    if (!marker.contains(idx)) {
                        sum += grid[y][x];
                    }
                }
            }
            return sum;
        }

        int result() {
            return sum_unmarked() * last_marked;
        }

        std::vector<std::vector<int>> grid{};
        std::unordered_set<int> marker{};
        bool completed;
        int last_marked;
    };

    struct Game {
        explicit Game(std::vector<Board> &boards) : boards{boards} {}

        int run(std::vector<int> &numbers, bool early_exit = true) {
            std::vector<int> results{};
            for (const auto &number: numbers) {
                for (auto &board: boards) {
                    if (board.completed) continue;

                    board.mark(number);
                    if (board.is_completed()) {
                        results.emplace_back(board.result());
                        if (early_exit)
                            return results.front();
                    }
                }
            }
            return results.back();
        }

        std::vector<Board> boards;
    };

    [[nodiscard]] auto get_numbers(const std::vector<std::string> &lines) {
        assert(!lines.empty());

        auto tokens = utils::split(lines[0], ',');
        auto result = std::vector<int>(tokens.size());
        std::transform(tokens.begin(), tokens.end(), result.begin(),
                       [](const std::string &line) {
                           return std::stoi(line);
                       });
        return result;
    }

    [[nodiscard]] auto create_board(const std::vector<std::string> &lines) {
        auto boards = std::vector<Board>{};
        auto tmp = Board{};

        for (int i = 1; i < lines.size(); ++i) {
            const auto &line = lines[i];
            const auto tokens = utils::split(line, ' ');
            auto numbers = std::vector<int>(tokens.size());
            std::transform(tokens.begin(), tokens.end(), numbers.begin(),
                           [](const std::string &line) {
                               return std::stoi(line);
                           });
            tmp.add_row(numbers);
            if ((i) % 5 == 0 && i > 1) {
                boards.emplace_back(tmp);
                tmp = Board{};
            }
        }
        return boards;
    }

    [[nodiscard]] int part1(const std::vector<std::string> &lines) {
        auto numbers = get_numbers(lines);
        auto boards = create_board(lines);
        return Game(boards).run(numbers);
    }

    [[nodiscard]] int part2(const std::vector<std::string> &lines) {
        auto numbers = get_numbers(lines);
        auto boards = create_board(lines);
        return Game(boards).run(numbers, false);
    }
}

File: 2021/rust/src/day4.rs
type Cell = (i32, bool);
type Grid = Vec<Vec<Cell>>;

struct Board {
    grid: Grid,
    last_value: i32,
    completed: bool,
}

impl Board {
    fn new(grid: Grid) -> Board {
        Board { grid, last_value: 0, completed: false }
    }

    fn mark(&mut self, value: i32) {
        for y in 0..5 {
            for x in 0..5 {
                if self.grid[y][x].0 == value {
                    self.grid[y][x] = (self.grid[y][x].0, true);
                    self.last_value = value;
                }
            }
        }
    }

    fn is_complete(&mut self) -> bool {
        for i in 0..5 {
            if (self.grid[0][i].1 && self.grid[1][i].1 && self.grid[2][i].1 && self.grid[3][i].1 && self.grid[4][i].1) ||
                (self.grid[i][0].1 && self.grid[i][1].1 && self.grid[i][2].1 && self.grid[i][3].1 && self.grid[i][4].1) {
                self.completed = true;
                return true;
            }
        }
        false
    }

    fn sum(&self) -> i32 {
        let mut s = 0;
        for y in 0..5 {
            for x in 0..5 {
                if !self.grid[y][x].1 {
                    s += self.grid[y][x].0;
                }
            }
        }
        s
    }

    fn result(&self) -> i32 {
        self.sum() * self.last_value
    }
}

struct Game<'a> {
    boards: &'a mut Vec<Board>,
}

impl Game<'_> {
    fn new(boards: &mut Vec<Board>) -> Game {
        Game {boards: boards}
    }

    fn run(&mut self, numbers: &Vec<i32>, early_exit: bool) -> i32 {
        let mut results = vec![];
        for number in numbers {
            for board in self.boards.iter_mut() {
                if board.completed {
                    continue;
                }
                board.mark(*number);
                if board.is_complete() {
                    results.push(board.result());
                    if early_exit {
                        return results.pop().unwrap();
                    }
                }
            }
        }
        return results.pop().unwrap();
    }
}

fn create_numbers(line: &str) -> Vec<i32> {
    line.split(',').map(|t| { t.parse::<i32>().unwrap() }).collect()
}

fn create_board(lines: &[String]) -> Grid {
    let mut rows = vec![];
    for line in lines {
        let mut row = vec![];
        for token in line.split_whitespace() {
            row.push((token.parse::<i32>().unwrap(), false));
        }
        rows.push(row);
    }
    rows
}

pub fn part1(lines: &Vec<String>) -> i32 {
    let n = create_numbers(lines.get(0).unwrap());
    let mut boards = vec![];
    let lines = &lines[1..];
    for i in (0..lines.len()).step_by(5) {
        boards.push(Board::new(create_board(&lines[i..i + 5])));
    }
    Game::new(&mut boards).run(&n, true)
}

pub fn part2(lines: &Vec<String>) -> i32 {
    let n = create_numbers(lines.get(0).unwrap());
    let mut boards = vec![];
    let lines = &lines[1..];
    for i in (0..lines.len()).step_by(5) {
        boards.push(Board::new(create_board(&lines[i..i + 5])));
    }
    Game::new(&mut boards).run(&n, false)
}